zkevm-circuits/src/mpt_circuit/account_leaf.rs

```diff
@@ -0,0 +1,589 @@
+                KeyData::store_defaults(cb, &ctx.memory[key_memory(is_s)]);
+                // Store the new parent
+                ParentData::store(
+                    cb,
+                    &ctx.memory[parent_memory(is_s)],
+                    storage_rlc[is_s.idx()].expr(),
+                    true.expr(),
+                    false.expr(),
+                    storage_rlc[is_s.idx()].expr(),
+                );
+            }
+            // Proof types
+            config.is_non_existing_account_proof = IsEqualGadget::construct(
+                &mut cb.base,
+                config.main_data.proof_type.expr(),
+                MPTProofType::AccountDoesNotExist.expr(),
+            );
+            config.is_account_delete_mod = IsEqualGadget::construct(
+                &mut cb.base,
+                config.main_data.proof_type.expr(),
+                MPTProofType::AccountDestructed.expr(),
+            );
+            config.is_nonce_mod = IsEqualGadget::construct(
+                &mut cb.base,
+                config.main_data.proof_type.expr(),
+                MPTProofType::NonceChanged.expr(),
+            );
+            config.is_balance_mod = IsEqualGadget::construct(
+                &mut cb.base,
+                config.main_data.proof_type.expr(),
+                MPTProofType::BalanceChanged.expr(),
+            );
+            config.is_storage_mod = IsEqualGadget::construct(
+                &mut cb.base,
+                config.main_data.proof_type.expr(),
+                MPTProofType::StorageChanged.expr(),
+            );
+            config.is_codehash_mod = IsEqualGadget::construct(
+                &mut cb.base,
+                config.main_data.proof_type.expr(),
+                MPTProofType::CodeHashExists.expr(),
+            );
+
+            // Drifted leaf handling
+            config.drifted = DriftedGadget::construct(
+                cb,
+                &config.parent_data,
+                &config.key_data,
+                &key_rlc,
+                &leaf_no_key_rlc,
+                &drifted_bytes,
+                &ctx.r,
+            );
+
+            // Wrong leaf handling
+            config.wrong = WrongGadget::construct(
+                cb,
+                a!(ctx.mpt_table.address_rlc),
+                config.is_non_existing_account_proof.expr(),
+                &config.rlp_key[true.idx()].key_value,
+                &key_rlc[true.idx()],
+                &wrong_bytes,
+                config.is_in_empty_trie[true.idx()].expr(),
+                config.key_data[true.idx()].clone(),
+                &ctx.r,
+            );
+            let is_non_existing_account = not::expr(config.wrong.is_key_equal.clone().expr());
+
+            // Anything following this node is below the account
+            // TODO(Brecht): For non-existing accounts it should be impossible to prove
+            // storage leaves unless it's also a non-existing proof?
```

Comment:
It's been quite a while since I wrote this so I'm trying to reconstruct why I thought that so I may have been wrong then and I most definitely could be wrong now. :)

- Good point on the address not matching!
- For placeholder leafs the account data passed in by the prover could be anything I think, so the storage root hash could be any value passed into the circuit by the prover no?

Do we need to support `StorageDoesNotExist` for non-existing accounts?

Reply:
When there is a `NonExistingAccount` proof, any `StorageChanged` lookup into `MPTTable` would fail (as `proof_type` is part of the table).

If the attacker would try to prepare a `StorageChanged` proof with a non-existing account, the address wouldn't match when there is a wrong leaf. But indeed I am not sure whether the proof would fail in the case of the placeholder leaf. I think we would need to add a constraint to allow the placeholder leaf only for the `NonExistingAccount` proof. I can add this in the PR I am preparing.

Regarding `StorageDoesNotExist` for non-existing accounts - I think we don't need it, but @ed255 would know better.